### ISSUE TITLE
How to implement a simple authentication?

### ISSUE THREAD

**Issue by Fram90** [DigDes/SoapCore#85]
Hello everyone.
I'm not good at WCF so maybe the question is something stupid, but I don't understand how to implement basic User-Password authentication stategy using SoapCore.
I have .Net Core app, have Soap endpoind provided via SoapCore. I figured out that I need to provide UserNamePasswordValidator for my binding. But since it's Net Core app and there is no web.config, I can't attach custom UserNamePasswordValidator to my binding using .config. It there any way to solve this?
Thanks.

**Comment by kotovaleksandr** [comment]
You should use specialized auth-middleware for ASP.NET Core or write own middleware. SOAP Core makes only SOAP serialization\deserialization

**Comment by phillip-haydon** [comment]
@kotovaleksandr I concur with you and wrote my own BasicAuth middleware since it's not much code.

However it would be nice to have a flag to say that a service uses a particular auth. I'm converting an old SOAP service to .net core and the only thing missing is the Policy with:

`<http:BasicAuthentication xmlns:http="http://schemas.microsoft.com/ws/06/2004/policy/http"/>`

Without the policy, SoapUI is being fussy and refuses to send the Basic Auth header with the request because it thinks the service doesn't need auth. (I got around this by using Postman instead where I can just send the header regardless)

**Comment by kotovaleksandr** [comment]
@phillip-haydon Whats differ SOAPCore against old SOAP service? This policy part of WSDL?

**Comment by phillip-haydon** [comment]
There's some slight differences, nothing major. SoapCore add's the service method parameter as a node in the payload. 

This is WCF
```xml
   <soap:Body>
      <pul:UpdateADN>
         <!--Optional:-->
         <pul:guid>?</pul:guid>
         <!--Optional:-->
         <pul:adn>?</pul:adn>
      </pul:UpdateADN>
   </soap:Body>
```

This is SoapCore, as you can see it adds `value` which is the parameter name. 
```xml
   <soapenv:Body>
      <pul:UpdateAdn>
         <!--Optional:-->
         <pul:value>
            <!--Optional:-->
            <pul:Guid>?</pul:Guid>
            <!--Optional:-->
            <pul:Adn>?</pul:Adn>
         </pul:value>
      </pul:UpdateAdn>
   </soapenv:Body>
```

The WSDL however.

(I've renamed things in here)

This is WCF


```xml
<?xml version="1.0" encoding="utf-8"?>
<wsdl:definitions name="bananaService" targetNamespace="http://company.com/bananaIntegration" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/" xmlns:soap12="http://schemas.xmlsoap.org/wsdl/soap12/" xmlns:tns="http://company.com/bananaIntegration" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsx="http://schemas.xmlsoap.org/ws/2004/09/mex" xmlns:wsap="http://schemas.xmlsoap.org/ws/2004/08/addressing/policy" xmlns:wsaw="http://www.w3.org/2006/05/addressing/wsdl" xmlns:msc="http://schemas.microsoft.com/ws/2005/12/wsdl/contract" xmlns:wsp="http://schemas.xmlsoap.org/ws/2004/09/policy" xmlns:wsa10="http://www.w3.org/2005/08/addressing" xmlns:wsam="http://www.w3.org/2007/05/addressing/metadata">
    <wsp:Policy wsu:Id="CustomBinding_IbananaService_policy">
        <wsp:ExactlyOne>
            <wsp:All>
                <http:BasicAuthentication xmlns:http="http://schemas.microsoft.com/ws/06/2004/policy/http"/>
                <sp:TransportBinding xmlns:sp="http://schemas.xmlsoap.org/ws/2005/07/securitypolicy">
                    <wsp:Policy>
                        <sp:TransportToken>
                            <wsp:Policy>
                                <sp:HttpsToken RequireClientCertificate="false"/>
                            </wsp:Policy>
                        </sp:TransportToken>
                        <sp:AlgorithmSuite>
                            <wsp:Policy>
                                <sp:Basic256/>
                            </wsp:Policy>
                        </sp:AlgorithmSuite>
                        <sp:Layout>
                            <wsp:Policy>
                                <sp:Strict/>
                            </wsp:Policy>
                        </sp:Layout>
                    </wsp:Policy>
                </sp:TransportBinding>
            </wsp:All>
        </wsp:ExactlyOne>
    </wsp:Policy>
    <wsdl:types>
        <xsd:schema targetNamespace="http://company.com/bananaIntegration/Imports">
            <xsd:import schemaLocation="https://db-banana.api.company.net/banana?xsd=xsd0" namespace="http://company.com/bananaIntegration"/>
            <xsd:import schemaLocation="https://db-banana.api.company.net/banana?xsd=xsd1" namespace="http://schemas.microsoft.com/2003/10/Serialization/"/>
        </xsd:schema>
    </wsdl:types>
    <wsdl:message name="UpdateAdn">
        <wsdl:part name="parameters" element="tns:UpdateADN"/>
    </wsdl:message>
    <wsdl:message name="UpdateAdnResponse">
        <wsdl:part name="parameters" element="tns:UpdateADNResponse"/>
    </wsdl:message>
    <wsdl:portType name="IbananaService">
        <wsdl:operation name="Update">
            <wsdl:input wsaw:Action="http://company.com/bananaIntegration/IbananaService/Update" name="UpdateAdn" message="tns:UpdateAdn"/>
            <wsdl:output wsaw:Action="http://company.com/bananaIntegration/IbananaService/UpdateResponse" name="UpdateAdnResponse" message="tns:UpdateAdnResponse"/>
        </wsdl:operation>
    </wsdl:portType>
    <wsdl:binding name="CustomBinding_IbananaService" type="tns:IbananaService">
        <wsp:PolicyReference URI="#CustomBinding_IbananaService_policy"/>
        <soap12:binding transport="http://schemas.xmlsoap.org/soap/http"/>
        <wsdl:operation name="Update">
            <soap12:operation soapAction="http://company.com/bananaIntegration/IbananaService/Update" style="document"/>
            <wsdl:input name="UpdateAdn">
                <soap12:body use="literal"/>
            </wsdl:input>
            <wsdl:output name="UpdateAdnResponse">
                <soap12:body use="literal"/>
            </wsdl:output>
        </wsdl:operation>
    </wsdl:binding>
    <wsdl:service name="bananaService">
        <wsdl:port name="CustomBinding_IbananaService" binding="tns:CustomBinding_IbananaService">
            <soap12:address location="https://db-banana.api.company.net/banana"/>
        </wsdl:port>
    </wsdl:service>
</wsdl:definitions>
```

This is SoapCore

```xml
<wsdl:definitions xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:msc="http://schemas.microsoft.com/ws/2005/12/wsdl/contract" xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/" xmlns:tns="http://company.com/bananaIntegration" xmlns:wsam="http://www.w3.org/2007/05/addressing/metadata" targetNamespace="http://company.com/bananaIntegration" name="IbananaService" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/">
    <wsdl:types>
        <xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" targetNamespace="http://company.com/bananaIntegration">
            <xs:import namespace="http://schemas.microsoft.com/2003/10/Serialization/Arrays"/>
            <xs:import namespace="http://schemas.datacontract.org/2004/07/System"/>
            <xs:element name="UpdateAdn">
                <xs:complexType>
                    <xs:sequence>
                        <xs:element minOccurs="0" maxOccurs="1" name="value" type="tns:UpdateAdn"/>
                    </xs:sequence>
                </xs:complexType>
            </xs:element>
            <xs:element name="UpdateAdnResponse">
                <xs:complexType>
                    <xs:sequence>
                        <xs:element minOccurs="0" maxOccurs="1" name="UpdateAdnResult" type="tns:UpdateAdnResponse"/>
                    </xs:sequence>
                </xs:complexType>
            </xs:element>
            <xs:complexType name="UpdateAdn">
                <xs:sequence>
                    <xs:element minOccurs="0" maxOccurs="1" name="Guid" type="xs:string"/>
                    <xs:element minOccurs="0" maxOccurs="1" name="Adn" type="xs:string"/>
                </xs:sequence>
            </xs:complexType>
            <xs:element name="UpdateAdn" nillable="true" type="tns:UpdateAdn"/>
            <xs:complexType name="UpdateAdnResponse">
                <xs:sequence>
                    <xs:element minOccurs="0" maxOccurs="1" name="UpdateAdnResult" type="tns:UpdateAdnResult"/>
                </xs:sequence>
            </xs:complexType>
            <xs:element name="UpdateAdnResponse" nillable="true" type="tns:UpdateAdnResponse"/>
            <xs:complexType name="UpdateAdnResult">
                <xs:sequence>
                    <xs:element minOccurs="1" maxOccurs="1" name="Success" type="xs:boolean"/>
                    <xs:element minOccurs="1" maxOccurs="1" name="ErrorCode" type="xs:int"/>
                    <xs:element minOccurs="0" maxOccurs="1" name="ErrorMessage" type="xs:string"/>
                </xs:sequence>
            </xs:complexType>
            <xs:element name="UpdateAdnResult" nillable="true" type="tns:UpdateAdnResult"/>
        </xs:schema>
    </wsdl:types>
    <wsdl:message name="IbananaService_UpdateAdn_InputMessage">
        <wsdl:part name="parameters" element="tns:UpdateAdn"/>
    </wsdl:message>
    <wsdl:message name="IbananaService_UpdateAdn_OutputMessage">
        <wsdl:part name="parameters" element="tns:UpdateAdnResponse"/>
    </wsdl:message>
    <wsdl:portType name="IbananaService">
        <wsdl:operation name="UpdateAdn">
            <wsdl:input message="tns:IbananaService_UpdateAdn_InputMessage"/>
            <wsdl:output message="tns:IbananaService_UpdateAdn_OutputMessage"/>
        </wsdl:operation>
    </wsdl:portType>
    <wsdl:binding name="BasicHttpBinding_IbananaService" type="tns:IbananaService">
        <soap:binding transport="http://schemas.xmlsoap.org/soap/http"/>
        <wsdl:operation name="UpdateAdn">
            <soap:operation soapAction="http://company.com/bananaIntegration/IbananaService/UpdateAdn" style="document"/>
            <wsdl:input>
                <soap:body use="literal"/>
            </wsdl:input>
            <wsdl:output>
                <soap:body use="literal"/>
            </wsdl:output>
        </wsdl:operation>
    </wsdl:binding>
    <wsdl:service name="IbananaService">
        <wsdl:port name="BasicHttpBinding_IbananaService" binding="tns:BasicHttpBinding_IbananaService">
            <soap:address location="https://db-banana-api.company.net/Service.asmx"/>
        </wsdl:port>
    </wsdl:service>
</wsdl:definitions>
```

**Comment by kotovaleksandr** [comment]
Thanks! Right, tag http:BasicAuthentication missed in SoapCore, whats attribute or setting set this tag to WCF?

**Comment by kotovaleksandr** [comment]
```xml
 <basicHttpBinding>
        <binding>
          <security mode="TransportCredentialOnly">
            <transport clientCredentialType="Basic"/>
          </security>
        </binding>
      </basicHttpBinding>
```
its your configuration?

**Comment by phillip-haydon** [comment]
I'm not 100% sure, looking at the old project now.

In the bindings I have.

```xml
      <customBinding>
        <binding name="basicAuth">
          <textMessageEncoding messageVersion="Soap12" />
          <httpsTransport authenticationScheme="Basic" />
        </binding>
      </customBinding>
```

The service is setup with:

```xml
      <service name="ThirdParty.Banana.BananaService">
        <endpoint address="" binding="customBinding" contract="ThirdParty.Banana.IPulsaService" bindingNamespace="http://company.com/bananaIntegration" bindingConfiguration="basicAuth">
        </endpoint>
      </service>
```

Then there's a custom implementation of the `IServiceHostFactory` interface which implements:

```csharp
        public void ApplyCredentials(ref ServiceHost host)
        {
            host.Credentials.UserNameAuthentication.UserNamePasswordValidationMode =
                UserNamePasswordValidationMode.Custom;
            host.Credentials.UserNameAuthentication.CustomUserNamePasswordValidator = CustomUserNamePasswordValidator();
        }
```

`CustomUserNamePasswordValidator` just implements `UserNamePasswordValidator.Validate` which accepts the username/password to be checked.

I would prefer to keep my own middleware to do the basic auth. Just would like to get the policy in somehow so that people getting the wsdl know that it requires auth.

**Comment by kotovaleksandr** [comment]
@phillip-haydon Please, check on my fork:
https://github.com/kotovaleksandr/SoapCore
You can enable basic auth on your service with this code:
```csharp
Binding binding = new BasicHttpBinding
{
	Security = new BasicHttpSecurity
	{
		Mode = BasicHttpSecurityMode.TransportCredentialOnly,
		Transport = new HttpTransportSecurity
		{
			ClientCredentialType = HttpClientCredentialType.Basic
		}
	}
};
app.UseSoapEndpoint<ISampleService>("/Service.svc", binding);
```

**Comment by phillip-haydon** [comment]
OH. So fast! <3

I will pull this down tomorrow and give it a go. Super busy today >< too much stress. (fighting with WCF not sending auth header)

**Comment by phillip-haydon** [comment]
OK I've tested it, I think there's 1 small thing missing which is:

```
    <wsdl:binding name="CustomBinding_IbananaService" type="tns:IbananaService">
        <wsp:PolicyReference URI="#CustomBinding_IbananaService_policy"/>
```

In the binding of the service it should reference the policy. 

So the policy is in place and that's correct, just need to include the reference to the policy in the binding and then it's perfecttttt

<3

**Comment by kotovaleksandr** [comment]
@phillip-haydon Check again

**Comment by phillip-haydon** [comment]
Sorry for the late reply. I just tried it now but there's no change? I pulled latest from your branch but in my test it's not including the reference.

**Comment by kotovaleksandr** [comment]
Try debug class MetaWCFBodyWriter, line 567: _binding.HasBasicAuth(). This code should add necessary policy to wsdl.

**Comment by kotovaleksandr** [comment]
@phillip-haydon Any news? If you can share your code to me - i can debug.

**Comment by phillip-haydon** [comment]
Just busy at work. I’ll have time tomorrow morning to retest. 

Sorry for delays :(

**Comment by phillip-haydon** [comment]
Hey,

Ok so I cannot get it to go into `MetaWCFBodyWriter` at all. I pulled the project into mine and debugged it, I step into `SoapEndpointExtensions` without issue but never into the body writer.

```
            var binding = new BasicHttpBinding
            {
                Security = new BasicHttpSecurity
                {
                    Mode = BasicHttpSecurityMode.TransportCredentialOnly, 
                    Transport = new HttpTransportSecurity
                    {
                        ClientCredentialType = HttpClientCredentialType.Basic
                    }
                }
            };
            
            app.UseSoapEndpoint<IPulsaService>("/Service.svc", binding,
                SoapSerializer.DataContractSerializer, caseInsensitivePath: true);
            app.UseSoapEndpoint<IPulsaService>("/Service.asmx", binding,
                SoapSerializer.XmlSerializer, caseInsensitivePath: true);

**Comment by phillip-haydon** [comment]
Hey,

Ok so I cannot get it to go into `MetaWCFBodyWriter` at all. I pulled the project into mine and debugged it, I step into `SoapEndpointExtensions` without issue but never into the body writer.

```
            services.AddSoapCore();
...
            var binding = new BasicHttpBinding
            {
                Security = new BasicHttpSecurity
                {
                    Mode = BasicHttpSecurityMode.TransportCredentialOnly, 
                    Transport = new HttpTransportSecurity
                    {
                        ClientCredentialType = HttpClientCredentialType.Basic
                    }
                }
            };
            
            app.UseSoapEndpoint<IBananaService>("/Service.svc", binding,
                SoapSerializer.DataContractSerializer, caseInsensitivePath: true);
            app.UseSoapEndpoint<IBananaService>("/Service.asmx", binding,
                SoapSerializer.XmlSerializer, caseInsensitivePath: true);
```

This is my setup config which is based on the wiki and the binding you gave me. 

I'm gonna try setup a new project after lunch so I can share something reproducible with you.

**Comment by kotovaleksandr** [comment]
I'm add binding policy for DataContractSerializer (WCF compatible WSDL), you check on XmlSerializer? I can add binding policy on next week for XmlSerializer.

**Comment by phillip-haydon** [comment]
Sorry for late reply, chaos, nothing but chaos. 

Yes I'm using the XmlSerializer because it's the closest I can get the XML to match. The old endpoint uses. 

- `MessageContract`
- `MessageBodyMember`

But if I use these then SoapCore fails to bind the incoming message. So I moved to the XmlSerializer to get it working. 

Using the `DataContractSerializer`.

My modal is actually using `DataContract` / `DataMember` but the `DataContractSerializer` fails to generate the schema so I just used the `XmlSerializer` for it to work.

**Comment by smg-bg** [comment]
I've used the example here https://jasonwatmore.com/post/2019/10/21/aspnet-core-3-basic-authentication-tutorial-with-example-api#users-controller-cs

However, even though I'm returning AuthenticateResult.Fail("Cannot read authorization header.") the request continues to and a result is return to the user with status 200 (instead of 401). I suspect it is something with the SoapEndpointMiddleware that is overriding the results from the AuthenticationHandler. 

I'm kinda stuck and cannot find the reason for this behavior. Any ideas?

**Comment by mtpultz** [comment]
@kotovaleksandr I'm performing a time boxed technical spike in a sprint to determine whether we can consume SoapCore as a dependency in our .Net Core application.  It was easy to setup and determine that we can have SOAP endpoints alongside our RESTful JSON API, but I'm trying determine whether SoapCore supports security settings through HTTP bindings.  The HTTP security bindings available in .Net Core suggests that it could support basic, digest, and certificate, but this ticket seems to contain the most information that I can find on implementation.  Are you able to point at other resources other than this open ticket regarding implementation, or whether this is a valid assumption regarding the HTTP bindings?

When I add the bindings I would expect that I wouldn't get a result on a POST from Postman with this payload, but I do as if the endpoint is anonymous:

```xml
<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
	  <SetModel xmlns="http://tempuri.org/">
		<id>7</id>
		<email>test@example.com</email>
	  </SetModel>
  </soap:Body>
</soap:Envelope>
```

Is there more documentation or examples of securing SOAP endpoints?  

```C#
Binding binding = new BasicHttpBinding
{
    Security = new BasicHttpSecurity
    {
        Mode = BasicHttpSecurityMode.Transport,
        Transport = new HttpTransportSecurity
        {
            ClientCredentialType = HttpClientCredentialType.Basic
        }
    }
};

app.UseEndpoints(endpoints =>
{
    endpoints.MapControllers();
    endpoints.UseSoapEndpoint<IExampleService>("/Service.asmx", binding, SoapSerializer.XmlSerializer);
});
```

**Comment by kotovaleksandr** [comment]
@mtpultz Martin, i should repeat: SoapCore doesn't support any security mechanism, its only HTTP handler for SOAP requests. BindingsExtensions word only for WSDL generation.
If you need any security (token, basic, certificate) you should use auth proxy like a nginx or write custom middleware or write custom ServiceOperationTuner for access to headers in request.
For example, my SOAP service use client certificate authentication. For this ahead of service work nginx with ssl policy, which set client certificate to header and pass them to my service. Service with ServiceOperationTuner get certificate from header and process request.

**Comment by robvangeloven** [comment]
> @mtpultz Martin, i should repeat: SoapCore doesn't support any security mechanism, its only HTTP handler for SOAP requests. BindingsExtensions word only for WSDL generation.
> If you need any security (token, basic, certificate) you should use auth proxy like a nginx or write custom middleware or write custom ServiceOperationTuner for access to headers in request.
> For example, my SOAP service use client certificate authentication. For this ahead of service work nginx with ssl policy, which set client certificate to header and pass them to my service. Service with ServiceOperationTuner get certificate from header and process request.

I tried following your recommendation, but in a .NET Core 3 app there's the following wireup:
```cs
var pipeline = routes
				.CreateApplicationBuilder()
				.UseMiddleware<SoapEndpointMiddleware<T_MESSAGE>>(options)
				.Build();

			return routes.Map(pattern, pipeline)
				.WithDisplayName("SoapCore");
```
in SoapEndpointExtensions.cs

This of course ignores any other middleware except SoapEndpointMiddleware. I have some POC code that would add middleware to this custom map, would you be interested in this?

**Comment by kotovaleksandr** [comment]
Hi @robvangeloven ! It's question about auth?
This is my configuration for certificate auth:
```csharp
public void ConfigureServices(IServiceCollection services)
{
	services.AddSoapCore();
	services.AddSoapServiceOperationTuner(new SetCertificateServiceTuner());
}

public void Configure(IApplicationBuilder app, IWebHostEnvironment env, IServiceProvider serviceProvider)
{
	app.UseSoapEndpoint<ISynchronizationService>("/ExecutiveWorkplace/SyncService/Certificate", new BasicHttpBinding(), SoapSerializer.XmlSerializer);
}
```

**Comment by robvangeloven** [comment]
Hi @kotovaleksandr !

Thanks for the quick reply :-)

It has, for me, to do with auth yes. But it also has to do with your remark to do auth via the .net core middleware. 
I cannot do challenge-response through the service tuner, setting the response is being ignored because it just loops through the handlers or perhaps I've missed something? 

But also because the soap core route cannot be extended with any middleware right now. So for instance setting an [Authorize] attribute doesn't work. Since the whole ecosystem is standardising on using middleware and soapcore "ignores" those that makes integrating soapcore a lot more difficult into the .net core infrastructure. So for the auth part I'd have to write my own security handlers. But same goes for basically every piece of middleware

**Comment by kotovaleksandr** [comment]
@robvangeloven SoapCore doesn't know about any auth or other middlewares and work like anyone middleware.
If you need challenge auth - you should write custom middleware which do it, like this (Negotiate auth):
```csharp
public async Task Invoke(HttpContext context)
{
	if (!context.User.Identity.IsAuthenticated)
	{
		await context.ChallengeAsync();
		return;
	}
}
```
This middleware should exec **before** SoapCore middleware.
And after, as a option, you can use ServiceMethodTuner for access to auth result from previous middleware.

**Comment by robvangeloven** [comment]
@kotovaleksandr I think I've found the workaround :-)

Although Soap Core ignores attributes (like "Authorize") you can actually configure it via the endpoint:
```cs
endpoints.UseSoapEndpoint<IService>(opt =>
                {
                    opt.Binding = new BasicHttpBinding
                    {
                        Security = new BasicHttpSecurity
                        {
                            Mode = BasicHttpSecurityMode.TransportCredentialOnly,
                            Transport = new HttpTransportSecurity
                            {
                                ClientCredentialType = HttpClientCredentialType.Basic
                            }
                        }
                    };
                    opt.Path = "/Service.asmx";
                    opt.SoapSerializer = SoapSerializer.XmlSerializer;
                })
                .RequireAuthorization();
            });
```

Perhaps something that can be added in the wiki? It seems that a few people had the same "issue" I had

**Comment by nicostubi** [comment]
Hello guys,

Testing SoapCore for a SOAP/XML and client certificate authentication scenario.

As I read above, "SoapCore doesn't support any security mechanism, it is only an HTTP handler for SOAP requests.

A few questions:

About "Write custom middleware to handle client certificates": are you aware of any existing and reliable implementation?
About "Custom Service Operation Tuner": would it be an idea to integrate it within SoapCore?

BR
Nicolas

**Comment by robvangeloven** [comment]
> Hello guys,
> 
> Testing SoapCore for a SOAP/XML and client certificate authentication scenario.
> 
> As I read above, "SoapCore doesn't support any security mechanism, it is only an HTTP handler for SOAP requests."
> 
> A few questions:
> 
> About "Write custom middleware to handle client certificates": are you aware of any existing and reliable implementation?
> About "Custom Service Operation Tuner": would it be an idea to integrate it within SoapCore?
> 
> BR
> Nicolas

I would look into just using the [Authorize] header and using either policies or a default policy using client certs.

And just use the one Microsoft provided: https://docs.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-3.1

**Comment by cetk** [comment]
I am using SOAP core with basic authentication. Service endpoint works fine with postman tool passing basic authorization with credentials but for SOAP UI work only after setting  Pre-emptive auth to Authenticate pre-emptively.
Is it possible to do it with global preferences ?
